### DISCUSSION TITLE
Add option to Action to opt-out of unsaved changes alerts

### DISCUSSION THREAD

**Discussion by juliangums** [filamentphp/filament#13438] [Ideas]
@danharrin When I use `->unsavedChangesAlerts()` on my panel I get unsaved changes alerts on all mounted actions as you've defined it here https://github.com/filamentphp/filament/blob/3.x/packages/panels/resources/views/components/unsaved-action-changes-alert.blade.php#L11

However, not all mounted actions include forms with changes. I think an option to opt-out of this for custom actions and also having `ViewAction` opted-out would be good. I have a view-only slide out I am using for example from this plugin https://filamentphp.com/plugins/ralphjsmit-activitylog-pro

**Comment by danharrin** [comment]
Go ahead if you want to PR `->unsavedChangesAlerts(false)` on the Action class, I think it would be tricky though as you need to know if the current action has that setting ahead of time before the `unload` event listener happens

**Reply by ralphjsmit** [reply]
I just looked into this and it seems to be quite tricky indeed. One way it could be done is by iterating over every possible action on initial page load, then collecting the action names that have a form, and then only show the alert for those action names. But there doesn't seem to be a great way to just simply collect all actions on a page (or based on the interface). 

An other thing I looked at is by doing it with an heuristic like checking whether the mounted actions have any mounted data in it (bit ugly code yet but I assume you get the gist):

```php
                        ...(@js($this instanceof \Filament\Actions\Contracts\HasActions) ? (($wire.mountedActionsData.filter((mountedActionData) => Object.keys(mountedActionData).length > 0).length > 0) ? ($wire.mountedActions ?? []) : []) : []),
```

This works OK, until I suspected/discovered that infolists also make use of the same property, and then it defeats a bit the whole idea.

So all in all I think it would be a pretty difficult/tricky thing indeed to get right, though it also feels like a bit of a bug (cause in effect, every action with a plain infolist would trigger the warning). I don't find it that disturbing personally, but it's a consideration. One could also regard the popup as a warning "hey, you still have a popup open" instead of strictly "hey, you only have unsaved changes" (bc if one has an action open, that would suggest one is in the midst of performing/checking something).

**Reply by danharrin** [reply]
Yeah I also had that idea when implementing the feature in the first place, "you were in the middle of deleting something but you haven't deleted it yet, are you sure you want to leave?"

**Reply by synsyst** [reply]
Perhaps we could have it made on page level for specific actions? I've run into the issue that the filament standard save and cancel buttons don't work when I have tables injected into form via livewires (a form of related table views) - so I've had to make custom actions on edit page, but this causes a lot of "unsavedchanges" alerts since it's set on panelprovider level - so I'm left with the choice of having users getting an alert even if they press save and close if that action has a url redirect in it (usually back to index), or not having it at all on the entire panel even when users navigate away from a form where are actual unsaved changes.

Ideas?

**Reply by ralphjsmit** [reply]
If you want it on page level then I suppose you can just disable it on `Filament::getCurrentPanel()->unsavedChangesAlert(false)` in the `mount()` method of a page. Only thing to check then would be if it persists across later requests or if you would need a persistent middleware.
